### PR TITLE
[mlir][scf]: Expose emitNormalizedLoopBounds/denormalizeInductionVariable util functions (NFC)

### PR DESCRIPTION
**Pull request by AviadCo**

### CODE REVIEW

File: mlir/include/mlir/Dialect/SCF/Utils/Utils.h
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/SCF/Utils/Utils.h:125]: I am OK with this, but i'd really like this to be `OpFoldResult`s.

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:486]: Why `auto`?

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:624]: you would need to use `getAsConstantIndexValue(rewriter, loc, newLoopParams.lowerBound)` here and below.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp
Comment (AviadCo) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp]: good eye, by mistake.

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp
Comment (AviadCo) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp]: Good point, used `getValueOrCreateConstantIndexOp` to avoid wrong cast if Attribute is in `newLoopParams.x`

**Review by AviadCo — COMMENTED** [review]
Thanks for the code review!

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:494]: Actually you can make this
```
OpFoldResult newLowerBound = isZeroBased ? lb : rewriter.getOneAttr(lb.getType());
```

Same for the lower bound and step.

**Review by MaheshRavishankar — APPROVED** [review]
One more comment, but looks good to me!

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp
Comment (AviadCo) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp:494]: Once I moved it to attribute, it relevealed that it may lead to this error:
`'scf.for' op failed to verify that all of {lowerBound, upperBound, step} have same type`

This is because we now use `getValueOrCreateConstantIndexOp`, and the lowerBound / upperBound type is i32, but the new step is created from attribute into index type.

The error from the lit test:

```
// -----// IR Dump After InterpreterPass Failed (transform-interpreter) //----- //
"builtin.module"() ({
  "func.func"() <{function_type = () -> (), sym_name = "coalesce_i32_loops"}> ({
    %3 = "arith.constant"() <{value = 0 : i32}> : () -> i32
    %4 = "arith.constant"() <{value = 128 : i32}> : () -> i32
    %5 = "arith.constant"() <{value = 2 : i32}> : () -> i32
    %6 = "arith.constant"() <{value = 64 : i32}> : () -> i32
    %7 = "arith.constant"() <{value = 64 : i32}> : () -> i32
    %8 = "arith.constant"() <{value = 1 : index}> : () -> index
    %9 = "arith.constant"() <{value = 32 : i32}> : () -> i32
    %10 = "arith.constant"() <{value = 1 : index}> : () -> index
    %11 = "arith.muli"(%7, %9) <{overflowFlags = #arith.overflow<none>}> : (i32, i32) -> i32
    "scf.for"(%3, %11, %8) ({
    ^bb0(%arg1: i32):
      %12 = "arith.remsi"(%arg1, %9) : (i32, i32) -> i32
      %13 = "arith.divsi"(%arg1, %9) : (i32, i32) -> i32
      %14 = "arith.muli"(%12, %5) <{overflowFlags = #arith.overflow<none>}> : (i32, i32) -> i32
      %15 = "arith.muli"(%13, %5) <{overflowFlags = #arith.overflow<none>}> : (i32, i32) -> i32
      %16 = "arith.addi"(%15, %14) <{overflowFlags = #arith.overflow<none>}> : (i32, i32) -> i32
      "scf.yield"() : () -> ()
    }) {coalesce} : (i32, i32, index) -> ()
    "func.return"() : () -> ()
  }) : () -> ()
  "builtin.module"() ({
    "transform.named_sequence"() <{arg_attrs = [{transform.readonly}], function_type = (!transform.any_op) -> (), sym_name = "__transform_main"}> ({
    ^bb0(%arg0: !transform.any_op):
      %0 = "transform.structured.match"(%arg0) <{op_attrs = {coalesce}, ops = ["scf.for"]}> : (!transform.any_op) -> !transform.any_op
      %1 = "transform.cast"(%0) : (!transform.any_op) -> !transform.op<"scf.for">
      %2 = "transform.loop.coalesce"(%1) : (!transform.op<"scf.for">) -> !transform.op<"scf.for">
      "transform.yield"() : () -> ()
    }) : () -> ()
  }) {transform.with_named_sequence} : () -> ()
}) : () -> ()
```

lit test is:

`mlir/test/Dialect/SCF/transform-ops.mlir` line 279

Comment (AviadCo) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp]: I added new function called `getValueOrCreateConstantOp` so the bounds type will be maintains during the normalization

**Comment by AviadCo** [comment]
@MaheshRavishankar can please take a look at the last revision? thanks!

File: mlir/include/mlir/Dialect/Arith/Utils/Utils.h
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/Arith/Utils/Utils.h:59]: Nit: I think this only works for `int` ops. So maybe call it `getValueOrCreateContantIntOp`?

File: mlir/include/mlir/Dialect/SCF/Utils/Utils.h
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/SCF/Utils/Utils.h:140]: Nit: You can also make the operands `OpFoldResults`. Should be a relatively easy change since `Value` implicitly converts to `OpFoldResult`.

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (kuhar) [inline mlir/lib/Dialect/Arith/Utils/Utils.cpp:105]:
```suggestion
  if (auto value = dyn_cast_if_present<Value>(ofr))
```

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (kuhar) [inline mlir/lib/Dialect/Arith/Utils/Utils.cpp:107]: Why do we need two nested casts?

Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Arith/Utils/Utils.cpp:107]: Because `OpFoldResult` is a mix of `Value` and `Attribute` and needs conversion to `IntegerAttr`. But you could use `cast<IntergerAttr>` and drop the subsequent assert

**Review by MaheshRavishankar — APPROVED** [review]
If changing the function to take `OpFoldResult` arguments works without issues, this looks good to me.

File: mlir/include/mlir/Dialect/SCF/Utils/Utils.h
Comment (AviadCo) [reply mlir/include/mlir/Dialect/SCF/Utils/Utils.h:140]: Ack

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (AviadCo) [reply mlir/lib/Dialect/Arith/Utils/Utils.cpp]: Ack

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (AviadCo) [reply mlir/lib/Dialect/Arith/Utils/Utils.cpp]: right, done.

**Comment by AviadCo** [comment]
> If changing the function to take `OpFoldResult` arguments works without issues, this looks good to me.

I refactored the func args, it added a little more wrapping functions but works the same.

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (kuhar) [inline mlir/lib/Dialect/Arith/Utils/Utils.cpp:107]: What if `dyn_cast_if_present` returns null? Should we just have a separate null check upfront?

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (kuhar) [inline mlir/lib/Dialect/Arith/Utils/Utils.cpp:116]: also here

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (kuhar) [inline mlir/lib/Dialect/Arith/Utils/Utils.cpp:308]: also here

Comment (kuhar) [inline mlir/lib/Dialect/Arith/Utils/Utils.cpp:308]:
```suggestion
  if (auto value = dyn_cast_if_present<Value>(ofr))
    return value.getType();
  auto attr = cast<IntegerAttr>(dyn_cast_if_present<Attribute>(ofr));
```

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp
Comment (kuhar) [inline mlir/lib/Dialect/SCF/Utils/Utils.cpp:505]: The ternary operators make this quite complex. Instead, what if we assign `ub` as the default value and put the `createOrFold` in `if`s?

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (AviadCo) [reply mlir/lib/Dialect/Arith/Utils/Utils.cpp]: ofr can be either Value or Attribute, why would it fail?

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (AviadCo) [reply mlir/lib/Dialect/Arith/Utils/Utils.cpp]: same answer

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (AviadCo) [reply mlir/lib/Dialect/Arith/Utils/Utils.cpp]: same answer

Comment (AviadCo) [reply mlir/lib/Dialect/Arith/Utils/Utils.cpp]: Ack

File: mlir/lib/Dialect/SCF/Utils/Utils.cpp
Comment (AviadCo) [reply mlir/lib/Dialect/SCF/Utils/Utils.cpp]: Ack

File: mlir/lib/Dialect/Arith/Utils/Utils.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Arith/Utils/Utils.cpp:305]: Nit: I think it might be better to rename this just `getType` ?

Comment (MaheshRavishankar) [reply mlir/lib/Dialect/Arith/Utils/Utils.cpp:305]: Actually looking at its uses, you dont need the `getIntType` at all. You can just fold this into `getValueOrCreateConstantIntOp`?

**Comment by AviadCo** [comment]
@MaheshRavishankar can u please take another look?

**Review by MaheshRavishankar — COMMENTED** [review]
Looks fine to me! I already approved. Are you wait for me to land this?

**Comment by MaheshRavishankar** [comment]
Actually I just realized that `LoopParams` here is same as `Range` [here](https://github.com/llvm/llvm-project/blob/43bd7ae65af40ff3378d5a0395a058ba834ad8dd/mlir/include/mlir/Dialect/Utils/StaticValueUtils.h#L33) . I should have pointed it out earlier. Fine to do as a follow up, but would appreciate just using `Range` instead of `LoopParams`. If not, I can do as follow up.

**Comment by AviadCo** [comment]
> Looks fine to me! I already approved. Are you wait for me to land this?

I thought I need reapproval after new changes, will land this one

**Comment by AviadCo** [comment]
> Actually I just realized that `LoopParams` here is same as `Range` [here](https://github.com/llvm/llvm-project/blob/43bd7ae65af40ff3378d5a0395a058ba834ad8dd/mlir/include/mlir/Dialect/Utils/StaticValueUtils.h#L33) . I should have pointed it out earlier. Fine to do as a follow up, but would appreciate just using `Range` instead of `LoopParams`. If not, I can do as follow up.

I will do the follow up PR it is o.k.

**Comment by AviadCo** [comment]
@MaheshRavishankar Please review the follow up PR: https://github.com/llvm/llvm-project/pull/95501
@kuhar thanks for the review here also :)

**Pull request merged**
